### PR TITLE
Torsion

### PR DESCRIPTION
**Pull request by markhalonen**
- [x] Verify descriptors are working
- [x] Refactor with static `SampleType` class to handle Sample extensions and Icons
- [ ] unit tests?

- Excel export works without modification, but it does have "Engineering Stress" label instead of just "Stress"

My refactor commit is pretty big and doesn't really have anything to do with torsion. If we want to look at that in a separate PR, we can just revert that commit and then cherry-pick it onto a separate branch.

### CODE REVIEW

File: Libraries/src/net/relinc/libraries/sample/TorsionSample.java [hunk 1]
```diff
@@ -52,7 +56,44 @@ public String getParametersForPopover(boolean selected2) {
 		return "";
 	}
 	
-	public double getMeanRadius()
+	public double getStrainRate(double reflectedStrain) {
+		return 4 * (this.getAverageDiameter() / 2) * this.getWavespeed() / (this.barSetup.IncidentBar.diameter / 2 * this.getLength()) * reflectedStrain;
+	}
+	
+	public double getLoad(double strain) {
+		return this.getTorque((1) * strain) / (this.getAverageDiameter() / 2.0);
```
Comment (bhalonen) [inline Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: no times strain here.

Comment (bhalonen) [reply Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: This works if the input is bar strain, this should be clearly stated.

Comment (bhalonen) [reply Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: #Important function
function polar_moment_inertia_bar(bar_radius)
    π*bar_radius^4/2
end
#Important function
function surface_strain_to_bar_torque(surface_strain,bar_radius,shear_modulus)
    torque_bar=(shear_modulus|>u"N/m^2")*polar_moment_inertia_bar(bar_radius)*surface_strain/bar_radius
end

File: Libraries/src/net/relinc/libraries/staticClasses/SPOperations.java [hunk 2]
```diff
@@ -616,6 +619,8 @@ public static int findFirstIndexGreaterThanValue(double[] data, double val){
 		double[] integral = new double[arraySize];
 		if(!(x.length == y.length && beginInclusive < y.length && endInclusive < y.length))
 			return integral; // zeroed array.
+		
+		// y[0] is never used... is that right?
```
Comment (markhalonen) [inline Libraries/src/net/relinc/libraries/staticClasses/SPOperations.java]: @bhalonen this make sense to you?

File: Libraries/src/net/relinc/libraries/sample/LoadDisplacementSampleResults.java [hunk 3]
```diff
@@ -221,29 +232,50 @@ else if(load == null){
...
+			//must be hopkinson bar sample to get engineering stress.
+			HopkinsonBarSample hopkinsonBarSample = (HopkinsonBarSample)sample;
+			return hopkinsonBarSample.getEngineeringStressFromForce(load);
+		}
+		
 	}
 
 	public double[] getTrueStrain() {
+		if(sample instanceof TorsionSample) {
+			return this.getEngineeringStrain();
```
Comment (markhalonen) [inline Libraries/src/net/relinc/libraries/sample/LoadDisplacementSampleResults.java]: Should True strain be disabled in the UI?

File: Libraries/src/net/relinc/libraries/sample/TorsionSample.java [hunk 4]
```diff
@@ -108,7 +108,7 @@ private static double getShearModulus(double youngsModulus, double poissonRatio)
 	
 	public double getShearModulus()
 	{
-		return getShearModulus(this.getYoungsModulus(), POISSON_RATIO);
+		return getShearModulus(this.barSetup.IncidentBar.youngsModulus, POISSON_RATIO);
```
Comment (bhalonen) [inline Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: ouch. annoying error

**Comment by markhalonen** [comment]
From Brent
> Need wavespeed to be from the bar, not the sample
> Need to be able to specify Poisson ratio in bar creation

File: Libraries/src/net/relinc/libraries/sample/TorsionSample.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		return torque / momentOfInertia * ((this.getOuterDiameter() + this.getInnerDiameter()) / 4);
+	}
+	
+	public double getStressFromLoad(double load) {
+		double torque = load * (this.getAverageDiameter() / 2.0);
+		return this.getStressFromTorque(torque);
+	}
+	
+	@Override
+	public double getWavespeed() {
```
Comment (markhalonen) [inline Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: @bhalonen This is the method you're talking about? This actually uses all parameters from the bar setup if you look at `getShearModulus()`. Perhaps this should be moved to the bar setup class?

File: Libraries/src/net/relinc/libraries/sample/TorsionSample.java [hunk 6]
```diff
@@ -133,7 +132,7 @@ private static double getShearModulus(double youngsModulus, double poissonRatio)
 	
 	public double getShearModulus()
```
Comment (bhalonen) [inline Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: getBarShearModulus at the very least. There may be a shear modulus of the sample that we care about at some point

File: Libraries/src/net/relinc/libraries/sample/TorsionSample.java (hunk 5 above)
Comment (markhalonen) [reply Libraries/src/net/relinc/libraries/sample/TorsionSample.java]: Ok, will move into bar setup

**Pull request merged**